File: permutations/permutations.py
from typing import List
import copy

class Solution:

    def permute(self, nums: List[int]) -> List[List[int]]:
        
        def backtrack(candidates: List[int], solution: List[int]):

            if not candidates:
                solutions.append(copy.copy(solution))
                return

            for ind, candidate in enumerate(candidates):
                solution.append(candidate)
                candidates_copy = copy.copy(candidates)
                candidates_copy.pop(ind)
                backtrack(candidates_copy, solution)
                solution.pop(-1)
        
        solutions = list()
        if nums:
            backtrack(candidates=nums, solution=[])

        return solutions


if __name__ == "__main__":
    print(Solution().permute([1, 2, 3]))
    print(Solution().permute([5, 4, 2, 6]))

File: backtracking/permutations/permutations.go
package permutations

func backTrack(solutions *[][]int, candidates []int, solution []int) {

	if len(candidates) == 0 {
		solutionCopy := make([]int, len(solution))
		copy(solutionCopy, solution)
		*solutions = append(*solutions, solutionCopy)
		return
	}

	for ind, candidate := range candidates {
		solution = append(solution, candidate)
		candidatesCopy := make([]int, len(candidates))
		copy(candidatesCopy, candidates)

		candidatesCopy = append(candidatesCopy[:ind], candidatesCopy[ind+1:]...)
		backTrack(solutions, candidatesCopy, solution)

		solution = solution[:len(solution)-1]
	}

}

// Permute ...
func Permute(nums []int) [][]int {

	var solutions [][]int

	if len(nums) != 0 {
		backTrack(&solutions, nums, []int{})
	}

	return solutions
}
